### DISCUSSION TITLE
Classify Object Behavior to Enhance the Safety of Autonomous Vehicles (Project 221)

### DISCUSSION THREAD

**Discussion by robertogl** [mathworks/MATLAB-Simulink-Challenge-Project-Hub#53] [Collaborate]
Contribute to the discussion by asking and/or answering questions, commenting, or sharing your ideas for solutions to [project 221](https://github.com/mathworks/MathWorks-Excellence-in-Innovation/tree/main/projects/Classify%20Object%20Behavior%20to%20Enhance%20the%20Safety%20of%20Autonomous%20Vehicles)

---
*:robot: An AI bot may reply to your comments with suggestions and guidance. It will not solve the project for you. Your comments may be processed by a third-party LLM. To direct a question to a specific person, tag them with @username.*

**Comment by agentic-inquisit** [comment]


**Reply by robertogl** [reply]
Hi @DHEEPAK29,

Have you defined parameters for risky and safe behaviors? In that case you should be able to label your data and then start the training part.

**Reply by robertogl** [reply]
Hi @DHEEPAK29 

Happy new year! Sorry for the late response, but I am finally back. 

I am not sure how many instances you have created but you would need enough to be able to train a NN model. 
Then you should use data from the simulated sensors to understand whether the behavior of external objects, like pedestrian or other vehicles, is safe or not so that you can label the data (for example from a simulated camera images, or lidar, or both) and start the training. To characterize the objects' behavior as safe or not safe you will need to have info about their motion, so output from a tracking algorithm might be needed. 

After you have trained the model you will have to test it in a new test scenario and and check its performances.

I hope this gives you some inputs to move your project forward. Please let me know if you need anything else.

**Reply by robertogl** [reply]
Hi @DHEEPAK29,

That is a very good point. I think using just the 2 labels, 'Safe' and 'Unsafe' it is a great start and they are good enough to decide whether continue via the planned trajectory/task or change maneuver, respectively. In case of Unsafe state, there might be different reactions as you correctly pointed out, but that can be decided based on the obstacle's type and/or motion. However, I think that is normally taken care of by a different layer, which includes object detection. Does it make sense?

**Comment by VoidXia** [comment]
Hi @robertogl , I am also trying to work on this project. I was curious to know if we need to manually create the dataset or if there is an existing one we can use. While I understand the idea of creating various driving scenarios and labeling them as safe or unsafe, I am concerned that manually creating a sufficient amount of data for the neural network model's training may prove to be challenging. Additionally, you had mentioned the need for a tracking algorithm. I was wondering if you had any suggestions for such an algorithm. It seems to me that the result of the project "Traffic Data Analysis for Modeling and Prediction of Traffic Scenarios" might be of relevance here. I would love to hear your thoughts. Thanks!

**Reply by robertogl** [reply]
Hi @VoidXia,

I totally understand your concern. I am not aware of any dataset out there but that does not mean that some may not exist already.
Regarding creating a new one from a simulation environment, we were thinking that that process could be, at least partially, automated, by changing the environment by adding or removing objects, changing motion, behavior, etc.

We have many tracking algorithms in our [Sensor Fusion and Tracking Toolbox](https://www.mathworks.com/products/sensor-fusion-and-tracking.html). A simple Kalman Filter could work for objects with linear motion models but we have more advanced [multi-object trackers](https://www.mathworks.com/help/fusion/multi-object-trackers.html) and many [examples for autonomous systems](https://www.mathworks.com/help/fusion/examples.html?category=tracking-for-autonomous-systems&s_tid=CRUX_topnav).
Perhaps you can start from one of these examples and adapt it to your application.

**Reply by Eladki** [reply]
Yeah, I agree Roberto. You can track using the Sensor Fusion and Tracking Toolbox and the links that he provided are where I would start.

**Comment by K-Yogesh-Mishra** [comment]
Hello @robertogl I have submitted the solution to the problem, when should i expect a response from MATLAB side.

**Reply by robertogl** [reply]
That is great! Thanks for your submission. I have received the submission form with the link. The review should not take us more than a week.

**Reply by K-Yogesh-Mishra** [reply]
Hello @robertogl , please tell me the reviews i haven't got any mail also about my submission reviews

**Reply by robertogl** [reply]
I am working on it. However, you are supposed to upload the actual code in a GitHub repository, not a zip file that contains it. Also, the repository has no documentation at all. Can you at least provide some information about what your code does and some guidance on how to run it? You can add all that information in the README.md file. Thanks.

**Reply by K-Yogesh-Mishra** [reply]
I have provided a link in the README.md files where i have uploaded my project which involves all the documentation and a yout tube video on how our project is performing, please refer to that.
Thanks.

**Reply by robertogl** [reply]
Got it, thanks. You can paste all that information in the README file. If you are familiar with markdown you can have a similar nice formatting. Also, no need to zip the files for GitHub.

**Reply by robertogl** [reply]
Oh I forgot...There is no license in your repository. Would you be willing to release your code under an open source license (2-clause BSD or MIT)?

**Reply by K-Yogesh-Mishra** [reply]
Yes i am ok to release it as open source license with MIT and i will upload all the info in readme file. Do you need any other help for evaluating my solution.

**Reply by K-Yogesh-Mishra** [reply]
@robertogl Is my github repo ok for now?

**Reply by robertogl** [reply]
I see that you currently have the 'Apache-2.0' license, can you change it to 2-clause BSD or MIT? 
Also, it would be great if you could also add pics and videos in the README file. 
Thanks!

**Comment by K-Yogesh-Mishra** [comment]
Fixed! I have also uploaded the link which contains video and image
explanation please look into it.

**Comment by JakubSchwenkbeck** [comment]
Hey :)
Im working on this project since roughly a month and a half and wanted to ask if anyone had some ideas for extension/improvement : [Repository](https://github.com/JakubSchwenkbeck/Behavior-Classification-For-AV) . Im kind of stuck by creating data/own labels, I thought about generating both of them with a program but haven't had the time to do so yet. 
Happy to hear some feedback!

**Comment by JakubSchwenkbeck** [comment]
Hey!
I now submitted my project, but I still would like to have conversations about this topic, feel free to start any kind of discussion :)

**Comment by Byleenroy300903** [comment]
Hello everyone! This is Byleen from Loyola ICAM College of Engineering and Technology. I will be working on the project "Classify Object Behavior to Enhance the Safety of Autonomous Vehicles" which will let me to enhance my existing knowledge of what I currently have about AI and its uses.
